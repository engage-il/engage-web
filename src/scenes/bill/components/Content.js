// @flow
import React, { Component } from 'react'
import { createFragmentContainer, graphql } from 'react-relay'
import moment from 'moment'
import { Actions } from './Actions'
import { Element } from './Element'
import { CopyLink } from 'shared/components'
import { stylesheet, colors, mixins } from 'shared/styles'
import type { Bill } from 'shared/types'

const { floor } = Math

let Content = class Content extends Component {
  props: {
    bill: Bill
  }

  // lifecycle
  render () {
    const { bill } = this.props

    const now = moment()
    const date = moment(bill.hearing.date)
    const hoursLeft = floor(date.diff(now, 'hours', true))

    return <div>
      <div {...rules.header}>
        <section>
          <h1>{bill.title}</h1>
          <h4>{bill.documentNumber}</h4>
        </section>
        <section>
          <CopyLink value={document.location.href} />
        </section>
      </div>
      <div {...rules.body}>
        <div {...rules.column}>
          <Element label='State Synopsis'>{bill.summary}</Element>
        </div>
        <div {...rules.column}>
          <Element label='Hearing Date'>
            <span>{date.calendar()}</span>
            {hoursLeft < 24 && <span {...rules.hoursLeft}>
              {`(${hoursLeft} hours left)`}
            </span>}
          </Element>
          <Element label='Committee'>{bill.committee.name}</Element>
        </div>
      </div>
      <Actions bill={bill} />
    </div>
  }
}

Content = createFragmentContainer(Content, graphql`
  fragment Content_bill on Bill {
    documentNumber
    title
    summary
    sponsorName
    hearing {
      date
    }
    committee {
      name
    }
    chamber {
      name
    }
    ...Actions_bill
  }
`)

const rules = stylesheet({
  header: {
    display: 'flex',
    justifyContent: 'space-between',
    marginBottom: 15,
    '> section': {
      display: 'flex',
      flexDirection: 'column'
    },
    ...mixins.mobile({
      flexDirection: 'column',
      '> section:first-child': {
        marginBottom: 10
      },
      '> h1': {
        marginBottom: 5
      }
    })
  },
  body: {
    ...mixins.borders.low(['bottom']),
    display: 'flex',
    flexBasis: 0,
    paddingBottom: 15,
    marginBottom: 15,
    ...mixins.mobile({
      flexDirection: 'column'
    })
  },
  column: {
    flex: 1,
    ':not(:last-of-type)': {
      marginRight: 15
    },
    '> *:not(:last-child)': {
      marginBottom: 10
    },
    ...mixins.mobile({
      marginBottom: 15,
      ':last-child': {
        marginBottom: 0
      }
    })
  },
  hoursLeft: {
    marginLeft: 5,
    color: colors.secondary
  }
})

export { Content }
