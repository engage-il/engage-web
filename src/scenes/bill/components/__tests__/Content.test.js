/* eslint-env jest */
import React from 'react'
import moment from 'moment'
import { shallow } from 'enzyme'
import { Content } from '../Content'

// subject
let subject
let bill

function loadSubject () {
  subject = shallow(<Content bill={bill} />).dive()
}

const element = {
  title: () => subject.find('h1'),
  number: () => subject.find('h4'),
  summary: () => subject.find('Element').at(0),
  date: () => subject.find('Element span').at(0),
  actions: () => subject.find('Actions'),
  hoursLeft: () => subject.find('Element span').at(1),
  committee: () => subject.find('Element').at(2)
}

// specs
beforeEach(() => {
  // TODO: build rosie.js factories
  bill = {
    documentNumber: 'HB1234',
    title: 'Foo',
    summary: 'A bill, fantastic',
    hearing: {
      date: moment().add(11.9, 'hours').utc()
    },
    committee: {
      name: 'Many Pointed Bills'
    }
  }
})

afterEach(() => {
  subject = null
})

describe('#render', () => {
  describe('normally', () => {
    beforeEach(loadSubject)

    it('shows the title', () => {
      expect(element.title()).toHaveText(bill.title)
    })

    it('shows the document number', () => {
      expect(element.number()).toHaveText(bill.documentNumber)
    })

    it('shows the summary', () => {
      expect(element.summary()).toHaveProp('children', bill.summary)
    })

    it('shows the hearing date', () => {
      expect(element.date()).toHaveText(moment(bill.hearing.date).calendar())
    })

    it('shows the hours left', () => {
      expect(element.hoursLeft()).toHaveText('(11 hours left)')
    })

    it('shows the committee name', () => {
      expect(element.committee()).toHaveProp('children', bill.committee.name)
    })

    it('shows the actions', () => {
      expect(element.actions()).toHaveProp('bill', bill)
    })
  })

  describe('when the date is over 23 hours away', () => {
    it('does not show the hours left', () => {
      bill.hearing.date = moment().add(24.1, 'hours').utc()
      loadSubject()
      expect(element.hoursLeft().get(0)).toBeFalsy()
    })
  })
})

describe('the relay container', () => {
  it('exists', () => {
    expect(Content.container.fragment).toBeTruthy()
  })
})
