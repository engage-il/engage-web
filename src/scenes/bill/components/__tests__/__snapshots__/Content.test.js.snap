// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render normally 1`] = `
<div>
  <div
    data-css-1wwvkuu=""
  >
    <section>
      <h1>
        Foo
      </h1>
      <h4>
        HB1234
      </h4>
    </section>
    <section>
      <CopyLink
        value="about:blank"
      />
    </section>
  </div>
  <div
    data-css-1bq0tws=""
  >
    <div
      data-css-1qww4bd=""
    >
      <Element
        label="State Synopsis"
      >
        A bill, fantastic
      </Element>
    </div>
    <div
      data-css-1qww4bd=""
    >
      <Element
        label="Hearing Date"
      >
        <span>
          Tomorrow at 3:35 AM
        </span>
        <span
          data-css-1fo47jf=""
        >
          (11 hours left)
        </span>
      </Element>
      <Element
        label="Committee"
      >
        Many Pointed Bills
      </Element>
    </div>
  </div>
  <Actions
    bill={
      Object {
        "committee": Object {
          "name": "Many Pointed Bills",
        },
        "documentNumber": "HB1234",
        "hearing": Object {
          "date": "2017-10-24T03:35:46.636Z",
        },
        "summary": "A bill, fantastic",
        "title": "Foo",
      }
    }
  />
</div>
`;

exports[`#render when the date is over 23 hours away 1`] = `
<div>
  <div
    data-css-1wwvkuu=""
  >
    <section>
      <h1>
        Foo
      </h1>
      <h4>
        HB1234
      </h4>
    </section>
    <section>
      <CopyLink
        value="about:blank"
      />
    </section>
  </div>
  <div
    data-css-1bq0tws=""
  >
    <div
      data-css-1qww4bd=""
    >
      <Element
        label="State Synopsis"
      >
        A bill, fantastic
      </Element>
    </div>
    <div
      data-css-1qww4bd=""
    >
      <Element
        label="Hearing Date"
      >
        <span>
          Tomorrow at 3:47 PM
        </span>
      </Element>
      <Element
        label="Committee"
      >
        Many Pointed Bills
      </Element>
    </div>
  </div>
  <Actions
    bill={
      Object {
        "committee": Object {
          "name": "Many Pointed Bills",
        },
        "documentNumber": "HB1234",
        "hearing": Object {
          "date": "2017-10-24T15:47:46.674Z",
        },
        "summary": "A bill, fantastic",
        "title": "Foo",
      }
    }
  />
</div>
`;
